File: server/routes/agent.js
// ============================================================
//  server/routes/agent.js  —  Core Agent Reasoning Loop Routes
// ============================================================

const express = require("express");
const fs = require("fs");
const { resolvePath, getCWD } = require("../lib/cwd");
const { askOllama } = require("../lib/ollama");
const { cleanJson } = require("../../shared/json");
const { SYSTEM_PROMPT, PERSONAS } = require("../lib/prompts");
const { TOOLS, runCommandStreaming } = require("../tools");
const { loadState, saveState, appendLog } = require("./state");
const { loadMemory } = require("./workspace");
const { MAX_STEPS, MAX_CTX_CHARS, MAX_HISTORY_MESSAGES } = require("../../shared/constants");

const router = express.Router();

const pendingHuman = new Map();      // runId → resolve
const pendingConfirms = new Map();   // confirmId → resolve

const NEEDS_CONFIRM = new Set(["delete_file"]);

async function toolAskHuman(question, runId) {
  return new Promise(resolve => {
    pendingHuman.set(runId, resolve);
    setTimeout(() => {
      pendingHuman.delete(runId);
      resolve("(no answer — timed out)");
    }, 120000);
  });
}

async function autoplan(task, model, send) {
  send("planning", { message: "🏗 Making a plan…" });
  try {
    const raw = await askOllama([
      {
        role: "system",
        content: `You are a planning assistant. Given a coding task, produce a brief step-by-step plan.
Response format — ONLY valid JSON:
{"plan": ["step 1", "step 2", ...], "needs_files": ["file1.js"], "estimated_steps": 5}`
      },
      { role: "user", content: task }
    ], model, 60000);
    const d = JSON.parse(cleanJson(raw));
    send("plan_ready", { plan: d.plan || [], needs_files: d.needs_files || [], estimated_steps: d.estimated_steps || 5 });
    return d;
  } catch {
    return null;
  }
}

async function summarizeContext(ctx, model) {
  if (ctx.length <= MAX_CTX_CHARS) return ctx;
  try {
    const raw = await askOllama([
      {
        role: "system",
        content: "You are a context summarizer. Compress the agent conversation history into a concise plain-text summary preserving all important facts, decisions, file paths, and code changes. Return only the summary text — no JSON, no markdown."
      },
      {
        role: "user",
        content: `Summarize this agent session context:\n\n${ctx.slice(0, 20000)}`
      }
    ], model, 60000);
    // Use raw text directly — no JSON parsing needed
    const summary = raw.trim() || ctx.slice(-4000);
    return `[SUMMARIZED CONTEXT]\n${summary}`;
  } catch {
    return ctx.slice(-4000);
  }
}

function trimContext(ctx, req) {
  if (ctx.length <= MAX_CTX_CHARS) return ctx;
  return `[Original task]\n${req}\n\n[...older steps trimmed...]\n\n${ctx.slice(-4000)}`;
}

/** Trim message history to prevent token explosion — keep system + last N messages */
function trimMessages(messages) {
  if (messages.length <= MAX_HISTORY_MESSAGES + 1) return messages;
  const system = messages[0];
  const rest = messages.slice(1);
  return [system, ...rest.slice(-MAX_HISTORY_MESSAGES)];
}

// ── ENDPOINTS ─────────────────────────────────────────────

router.post("/human", (req, res) => {
  const { runId, answer } = req.body;
  const resolve = pendingHuman.get(runId);
  if (resolve) {
    pendingHuman.delete(runId);
    resolve(answer || "");
  }
  res.json({ ok: true });
});

router.post("/confirm", (req, res) => {
  const { runId, confirmed } = req.body;
  const resolve = pendingConfirms.get(runId);
  if (resolve) {
    pendingConfirms.delete(runId);
    resolve(confirmed);
  }
  res.json({ ok: true });
});

router.post("/run", async (req, res) => {
  const { message, model, auto_plan, workspace: activeWorkspace } = req.body;
  if (!message) return res.status(400).json({ error: "No message" });

  res.setHeader("Content-Type", "text/event-stream");
  res.setHeader("Cache-Control", "no-cache");
  res.setHeader("Connection", "keep-alive");
  res.flushHeaders();

  let aborted = false;
  req.on("close", () => { aborted = true; });

  const send = (type, data) => {
    if (aborted) return;
    try {
      res.write(`data: ${JSON.stringify({ type, ...data })}\n\n`);
    } catch { aborted = true; }
  };

  const { context, history } = loadState();
  const activeModel = model || process.env.AGENT_MODEL || "qwen3-coder-next:cloud";
  let ctx = context;

  let userMsg = message;
  if (message.toLowerCase() === "continue" && context) {
    userMsg = "Continue the previous task from where you left off. Review the context and proceed.";
  }

  // Auto-plan if requested
  if (auto_plan && message.length > 20) {
    await autoplan(message, activeModel, send);
  }

  // Summarize context if too large
  if (ctx.length > MAX_CTX_CHARS * 1.5) {
    send("summarizing", { message: "📝 Summarizing context…" });
    ctx = await summarizeContext(ctx, activeModel);
  }

  // Inject memory if workspace is active
  let memNote = "";
  if (activeWorkspace) {
    const mem = loadMemory();
    const wMem = mem.workspaces[activeWorkspace] || {};
    const gMem = mem.global || {};
    const globalStr = Object.keys(gMem).length ? `\n[GLOBAL MEMORY]\n${Object.entries(gMem).map(([k,v])=>`${k}: ${v}`).join('\n')}` : '';
    const wsStr = Object.keys(wMem).length ? `\n[WORKSPACE MEMORY: ${activeWorkspace}]\n${Object.entries(wMem).map(([k,v])=>`${k}: ${v}`).join('\n')}` : '';
    memNote = `${globalStr}${wsStr}`;
  }

  // System Prompt details
  const persona = req.body.persona ? PERSONAS[req.body.persona] : null;
  const personaNote = persona ? `\n\n## Active Persona: ${persona.name}\n${persona.focus}` : "";
  const cwdNote = `\n\n## Current Working Directory\n${getCWD()}`;
  const memoryNote = memNote ? `\n\n## Memory Context${memNote}` : "";
  const resume = ctx ? `\n\n## Resumed Session Context\n${ctx}` : "";

  // Cap loaded history to prevent token explosion from previous sessions
  const cappedHistory = history.slice(-MAX_HISTORY_MESSAGES);

  const messages = [
    { role: "system", content: SYSTEM_PROMPT + personaNote + cwdNote + memoryNote + resume },
    ...cappedHistory,
    { role: "user", content: userMsg }
  ];

  send("start", { message: "Agent started", model: activeModel });

  const fileSnapshots = {};
  let finalAnswer = null;

  try {
    for (let n = 1; n <= MAX_STEPS; n++) {
      if (aborted) break;
      send("step", { step: n, message: `Step ${n} — Thinking…` });

      let raw;
      try {
        raw = await askOllama(messages, activeModel);
      } catch (e) {
        send("error", { message: `Ollama error: ${e.message}` });
        break;
      }

      let data;
      try {
        const cleaned = cleanJson(raw);
        data = JSON.parse(cleaned);
        if (data.__plain__) {
          send("final", { message: data.text });
          messages.push({ role: "assistant", content: raw });
          ctx += `\n\n[Step ${n}] Final(plain): ${data.text}`;
          saveState(ctx, trimMessages(messages).slice(1));
          finalAnswer = data.text;
          break;
        }
      } catch {
        send("final", { message: raw });
        messages.push({ role: "assistant", content: raw });
        ctx += `\n\n[Step ${n}] Final(raw): ${raw.slice(0, 500)}`;
        saveState(ctx, trimMessages(messages).slice(1));
        finalAnswer = raw;
        break;
      }

      if (data.thought) {
        send("thought", { message: data.thought });
      }

      if (data.final !== undefined) {
        send("final", { message: data.final });
        messages.push({ role: "assistant", content: raw });
        ctx += `\n\n[Step ${n}] Final: ${data.final}`;
        saveState(ctx, trimMessages(messages).slice(1));
        appendLog({ step: n, type: "final", message: data.final });
        finalAnswer = data.final;
        break;
      }

      const toolName = data.tool;
      const args = data.args || {};
      if (!toolName) {
        messages.push({ role: "assistant", content: raw });
        // Trim messages in place to prevent explosion
        const trimmed = trimMessages(messages);
        messages.length = 0;
        messages.push(...trimmed);
        continue;
      }

      // ask_human
      if (toolName === "ask_human") {
        const question = args.question || args.q || "Can you clarify?";
        const askRunId = `ask_${Date.now()}`;
        send("ask_human", { question, runId: askRunId });
        const answer = await toolAskHuman(question, askRunId);
        send("human_answered", { answer });
        messages.push({ role: "assistant", content: raw });
        messages.push({ role: "user", content: `[Human answered]: ${answer}` });
        ctx += `\n\n[Step ${n}] Asked: ${question}\nAnswer: ${answer}`;
        continue;
      }

      // confirm needed
      if (NEEDS_CONFIRM.has(toolName)) {
        const preview = args.path || "";
        const confirmId = `confirm_${Date.now()}_${Math.random().toString(36).slice(2)}`;
        send("confirm_request", { tool: toolName, preview, step: n, runId: confirmId });
        const confirmed = await new Promise(resolve => {
          pendingConfirms.set(confirmId, resolve);
          setTimeout(() => {
            pendingConfirms.delete(confirmId);
            resolve(false);
          }, 90000);
        });
        if (!confirmed) {
          const result = "Cancelled by user.";
          send("tool_result", { tool: toolName, result });
          messages.push({ role: "assistant", content: raw });
          messages.push({ role: "user", content: `[Tool result: ${toolName}]\n${result}` });
          ctx += `\n\n[Step ${n}] Tool:${toolName} → Cancelled`;
          continue;
        }
      }

      // file snapshot before write
      if ((toolName === "write_file" || toolName === "replace_text") && args.path) {
        const absPath = resolvePath(args.path);
        if (!fileSnapshots[absPath]) {
          try {
            fileSnapshots[absPath] = fs.readFileSync(absPath, "utf8");
          } catch {
            fileSnapshots[absPath] = "";
          }
        }
      }

      send("tool_call", { tool: toolName, args });

      // Run tool
      let result;
      if (toolName === "run_command" && args.command) {
        send("stream_start", { tool: toolName, command: args.command });
        result = await runCommandStreaming(args.command, send);
        send("stream_end", { tool: toolName });
      } else {
        const fn = TOOLS[toolName];
        result = fn ? String(await Promise.resolve(fn(args))) : `ERROR: unknown tool '${toolName}'`;
      }

      // inline diff after write
      if ((toolName === "write_file" || toolName === "replace_text") && args.path) {
        const absPath = resolvePath(args.path);
        const before = fileSnapshots[absPath] || "";
        let after = "";
        try {
          after = fs.readFileSync(absPath, "utf8");
        } catch {}
        if (before !== after) {
          send("inline_diff", { path: args.path, before, after });
        }
      }

      send("tool_result", { tool: toolName, result: result.slice(0, 1000) });
      appendLog({ step: n, tool: toolName, args, result: result.slice(0, 500) });

      messages.push({ role: "assistant", content: raw });
      messages.push({ role: "user", content: `[Tool result: ${toolName}]\n${result}` });
      ctx += `\n\n[Step ${n}] Tool:${toolName} Result:${result.slice(0, 500)}`;
      ctx = trimContext(ctx, userMsg);

      // Trim messages array to prevent context explosion
      const trimmed = trimMessages(messages);
      messages.length = 0;
      messages.push(...trimmed);
    }
  } catch (e) {
    send("error", { message: e.message });
  }

  if (!finalAnswer) {
    send("paused", { message: `Reached ${MAX_STEPS} steps. Type 'continue' to resume.` });
    saveState(ctx, trimMessages(messages).slice(1));
  }

  res.end();
});

module.exports = router;

File: cli/agent.py
# ============================================================
#  cli/agent.py  —  CLI Agent Main Loop
# ============================================================

import os
import re
import sys
import json
import shutil
from pathlib import Path
from datetime import datetime

from cli.ui import clr, C, info, ok, warn, err, thought, tool_ev, step_ev, result_ev, print_banner
from cli.tools import (
    CWD, _bg_procs, execute_tool, list_processes, kill_process,
    cd, screenshot, UPLOAD_DIR, resolve
)

try:
    import requests
except ImportError:
    import subprocess
    subprocess.run([sys.executable, "-m", "pip", "install", "requests", "--break-system-packages", "-q"])
    import requests


# ─── STATE FOR CANCELLATION ──────────────────────────────────
class CancelledError(Exception):
    pass

# ─── CONFIG ───────────────────────────────────────────────────
OLLAMA_URL = os.getenv("OLLAMA_URL", "http://localhost:11434/api/chat")
MODEL = os.getenv("AGENT_MODEL", "qwen3-coder-next:cloud")
MAX_STEPS = int(os.getenv("MAX_STEPS", "30"))
CONTEXT_FILE = ".agent_state.json"
LOG_FILE = ".agent_log.jsonl"
MAX_CTX_CHARS = 14_000
MAX_HISTORY_MSGS = 30

SYSTEM_PROMPT = """You are NOVA — an advanced AI coding agent (2026). Senior full-stack engineer.

## Available Tools

### Filesystem
- read_file(path) — read file with line numbers
- read_lines(path, start, end) — read specific line range — USE FOR LARGE FILES
- write_file(path, content) — create or overwrite file
- append_file(path, content) — append to file
- replace_text(path, old, new) — targeted find-and-replace
- list_files(path) — list directory (call at most ONCE per directory)
- search_in_files(pattern, directory) — search text across files
- grep(pattern, path) — regex search
- create_dir(path) — create directory
- delete_file(path) — delete file/dir

### Shell & Code
- run_command(command) — execute shell command
- python_eval(code) — run Python code

### Git
- git_status() — git status
- git_diff(file?) — git diff

### Web & APIs
- http_get(url) — fetch URL content
- http_post(url, body, method, headers) — POST/PUT/PATCH to any API with JSON body
- search_web(query) — search DuckDuckGo

### File Discovery & Archives
- find_files(pattern, directory, ext, maxdepth) — find files by name or extension
- zip(action, archive, source, dest) — create/extract/list zip or tar.gz archives
- diff_files(file1, file2) — unified diff between any two files

### Code Quality
- lint(file, linter) — run ESLint (JS), flake8 (Python), or syntax check

### Other
- screenshot(url_or_path) — take screenshot
- cd(path) — change directory
- ask_human(question) — ask user for clarification
- think(thought) — plan complex tasks before acting

## Response Format
ALWAYS respond with ONLY a single valid JSON object. NO markdown, NO extra text.

Use tool: {"thought":"<why>","tool":"<name>","args":{...}}
Finished: {"thought":"<summary>","final":"<message to user>"}

## CRITICAL JSON Rules
- Output ONLY the JSON object — nothing before or after
- Never wrap in ```json or any markdown fences
- Escape all special characters: newlines → \\n, tabs → \\t, quotes → \\"

## Efficiency Rules — MUST FOLLOW
1. NEVER call list_files on the same directory more than once
2. Use read_lines for large files — NOT read_file repeatedly
3. Use think to plan complex multi-step tasks before starting
4. Don't re-read files you've already read
5. Once work is verified, return {"final":"..."} immediately — do NOT loop

## Engineering Standards (2026)
- Always read_file before editing
- Use replace_text for small changes, write_file for new files
- After writing code, verify with run_command
- Use ask_human when requirements are unclear
- Never store secrets in code — use environment variables
- Handle ALL error cases explicitly"""

# ─── STATE MANAGEMENT ─────────────────────────────────────────
def save_state(ctx, hist):
    capped = hist[-MAX_HISTORY_MSGS:] if len(hist) > MAX_HISTORY_MSGS else hist
    data = {"saved_at": datetime.now().isoformat(), "context": ctx, "history": capped}
    Path(CONTEXT_FILE).write_text(json.dumps(data, ensure_ascii=False, indent=2))

def load_state():
    if Path(CONTEXT_FILE).exists():
        try:
            d = json.loads(Path(CONTEXT_FILE).read_text())
            return d.get("context", ""), d.get("history", [])
        except:
            pass
    return "", []

def clear_state():
    for f in [CONTEXT_FILE, LOG_FILE]:
        if Path(f).exists():
            Path(f).unlink()

def append_log(e):
    try:
        with open(LOG_FILE, "a") as f:
            f.write(json.dumps(e, ensure_ascii=False) + "\n")
    except:
        pass

# ─── ROBUST JSON PARSER ──────────────────────────────────────
def clean_json(raw):
    if not raw: return "{}"
    try:
        json.loads(raw)
        return raw
    except:
        pass
    # Strip fences
    s = re.sub(r'^```(?:json)?\s*', '', raw, flags=re.M)
    s = re.sub(r'\s*```\s*$', '', s, flags=re.M).strip()
    try:
        json.loads(s)
        return s
    except:
        pass
    # Extract first { ... } with brace matching
    depth = 0; start = -1; in_str = False; esc = False
    for i, c in enumerate(raw):
        if esc: esc = False; continue
        if c == '\\' and in_str: esc = True; continue
        if c == '"': in_str = not in_str; continue
        if in_str: continue
        if c == '{':
            if depth == 0: start = i
            depth += 1
        elif c == '}':
            depth -= 1
            if depth == 0 and start != -1:
                candidate = raw[start:i+1]
                try:
                    json.loads(candidate)
                    return candidate
                except:
                    pass
    # Manual extraction fallback
    t = re.search(r'"thought"\s*:\s*"((?:[^"\\]|\\.)*)"', raw)
    tl = re.search(r'"tool"\s*:\s*"([^"]+)"', raw)
    fn = re.search(r'"final"\s*:\s*"((?:[^"\\]|\\.)*)"', raw)
    thought_v = t.group(1) if t else ""
    if tl:
        return json.dumps({"thought": thought_v, "tool": tl.group(1), "args": {}})
    if fn:
        return json.dumps({"thought": thought_v, "final": fn.group(1)})
    return json.dumps({"__plain__": True, "text": raw})

# ─── OLLAMA CALL ──────────────────────────────────────────────
def ask_ollama(messages):
    try:
        r = requests.post(OLLAMA_URL, json={
            "model": MODEL,
            "messages": messages,
            "stream": False,
            "options": {"temperature": 0.2}
        }, timeout=300)
        r.raise_for_status()
        return r.json()["message"]["content"]
    except Exception as e:
        return json.dumps({"thought": "API error", "final": f"Ollama error: {e}"})

# ─── AGENT LOOP ───────────────────────────────────────────────
def run_agent(user_request, context, chat_history, image_paths=None):
    cwd_note = f"\n\n## Current Working Directory\n{CWD[0]}"
    resume = f"\n\n## Resumed Session Context\n{context}" if context else ""
    sys_p = SYSTEM_PROMPT + cwd_note + resume

    # Cap history to prevent token explosion
    capped_history = chat_history[-MAX_HISTORY_MSGS:] if len(chat_history) > MAX_HISTORY_MSGS else chat_history

    messages = [{"role": "system", "content": sys_p}] + capped_history

    user_content = user_request
    if image_paths:
        img_list = ", ".join(str(p) for p in image_paths)
        user_content += f"\n\n[Attached images: {img_list}]"

    messages.append({"role": "user", "content": user_content})

    ctx = context
    for n in range(1, MAX_STEPS + 1):
        try:
            step_ev(n)
            raw = ask_ollama(messages)
            cleaned = clean_json(raw)
            try:
                data = json.loads(cleaned)
            except:
                ok("Plain text response.")
                messages.append({"role": "assistant", "content": raw})
                ctx += f"\n\n[Step {n}] Final: {raw[:500]}"
                save_state(ctx, messages[1:])
                return raw, ctx, messages[1:]

            if data.get("__plain__"):
                ok("Response.")
                print(f"\n  {data['text'][:500]}\n")
                messages.append({"role": "assistant", "content": raw})
                ctx += f"\n\n[Step {n}] Final: {data['text'][:300]}"
                save_state(ctx, messages[1:])
                return data["text"], ctx, messages[1:]

            if data.get("thought"):
                thought(data["thought"])

            if "final" in data:
                ok("Done.")
                messages.append({"role": "assistant", "content": raw})
                ctx += f"\n\n[Step {n}] Final: {data['final'][:300]}"
                save_state(ctx, messages[1:])
                append_log({"step": n, "final": data["final"]})
                return data["final"], ctx, messages[1:]

            tool_name = data.get("tool")
            args = data.get("args", {})
            if not tool_name:
                messages.append({"role": "assistant", "content": raw})
                continue

            # Handle "none" tool — model thinks no tool is needed but didn't use {"final":"..."}
            NONE_TOOL_NAMES = {"none", "null", "n/a", "no_tool", "no tool", "finish", "done", "respond"}
            if str(tool_name).lower().strip() in NONE_TOOL_NAMES:
                messages.append({"role": "assistant", "content": raw})
                messages.append({"role": "user", "content":
                    'You do not need a tool for this. Respond with ONLY: {"thought":"<brief>","final":"<your answer>"}'
                })
                continue

            # think tool: show thought but don't print result to user
            if tool_name == "think":
                t_val = args.get("thought", args.get("reasoning", ""))
                thought(f"[Planning] {t_val}")
                messages.append({"role": "assistant", "content": raw})
                messages.append({"role": "user", "content": f"[Tool result: think]\nThought recorded. Proceed with your plan."})
                ctx += f"\n\n[Step {n}] think: {t_val[:200]}"
                continue

            tool_ev(tool_name, args)
            result = execute_tool(tool_name, args)
            result_ev(result)
            messages.append({"role": "assistant", "content": raw})
            messages.append({"role": "user", "content": f"[Tool result: {tool_name}]\n{result}"})
            ctx += f"\n\n[Step {n}] {tool_name}: {str(result)[:300]}"
            if len(ctx) > MAX_CTX_CHARS:
                ctx = f"[Original task]\n{user_request}\n\n[...trimmed...]\n\n{ctx[-4000:]}"
            append_log({"step": n, "tool": tool_name, "args": args, "result": str(result)[:300]})

        except (KeyboardInterrupt, EOFError):
            warn("Task cancelled.")
            save_state(ctx, messages[1:])
            raise CancelledError()

    warn(f"Reached {MAX_STEPS} steps. State saved. Type 'continue' to resume.")
    save_state(ctx, messages[1:])
    return "PAUSED", ctx, messages[1:]

# ─── ATTACH IMAGE ─────────────────────────────────────────────
def attach_image(path):
    abs_p = Path(resolve(path))
    if not abs_p.exists():
        err(f"File not found: {path}")
        return None
    dest = UPLOAD_DIR / f"cli_{int(datetime.now().timestamp())}_{abs_p.name}"
    shutil.copy(abs_p, dest)
    ok(f"Attached: {dest}")
    return dest

# ─── MAIN SHELL LOOP ──────────────────────────────────────────
def main():
    global MODEL
    print_banner()
    context, chat_history = load_state()
    if context:
        warn(f"Resumed session. Type 'continue' to pick up.")
    else:
        info(f"CWD: {CWD[0]}")
    print(clr(C.DIM, "  /help for commands\n"))

    pending_images = []

    # One-shot mode: python -m cli "task"
    if len(sys.argv) > 1:
        task = " ".join(sys.argv[1:])
        try:
            answer, context, chat_history = run_agent(task, context, chat_history)
        except CancelledError:
            err("Task cancelled.")
            return
        if answer != "PAUSED":
            print(f"\n{clr(C.GREEN + C.BOLD, '  NOVA:')}\n  {answer}\n")
        return

    while True:
        try:
            prompt = clr(C.BOLD + C.WHITE, f"[{Path(CWD[0]).name}]>>> ")
            user = input(prompt).strip()
        except EOFError:
            print()
            ok("Bye!")
            break
        except KeyboardInterrupt:
            print()
            continue
        if not user:
            continue

        lo = user.lower()

        if lo in ("exit", "quit"):
            ok("Bye!")
            break

        if lo == "/help":
            print(f"""
  {clr(C.BOLD, 'Commands:')}
  Ctrl+C             Cancel current task
  Ctrl+D / exit      Exit the CLI
  /clear             Clear session state
  /status            Show session info
  /log               Show last 10 log entries
  /model <name>      Switch Ollama model
  /tools             List all available tools
  /ps                Show background processes
  /kill <id>         Kill a background process
  /attach <path>     Attach image to next message
  /screenshot <url>  Take a screenshot
  /save [filename]   Export session to markdown file
  /cwd               Show current directory
  /cd <path>         Change directory
  continue           Resume a paused session
""")
            continue

        if lo == "/clear":
            clear_state()
            context, chat_history = "", []
            pending_images = []
            ok("Session cleared.")
            continue

        if lo == "/status":
            info(f"Model: {MODEL}")
            info(f"CWD: {CWD[0]}")
            info(f"Context: {len(context)} chars")
            info(f"History: {len(chat_history)} messages")
            if pending_images:
                info(f"Pending images: {len(pending_images)}")
            continue

        if lo == "/ps":
            list_processes()
            continue

        if lo.startswith("/kill "):
            try:
                kill_process(int(lo.split()[1]))
            except:
                err("Usage: /kill <id>")
            continue

        if lo.startswith("/model "):
            MODEL = user[7:].strip()
            ok(f"Model: {MODEL}")
            continue

        if lo == "/tools":
            from cli.tools import TOOL_MAP
            print("\n  " + "\n  ".join(f"• {t}" for t in sorted(TOOL_MAP.keys())) + "\n")
            continue

        if lo == "/cwd":
            info(f"CWD: {CWD[0]}")
            continue

        if lo.startswith("/cd "):
            cd(user[4:].strip())
            continue

        if lo.startswith("/attach "):
            img = attach_image(user[8:].strip())
            if img:
                pending_images.append(img)
            continue

        if lo.startswith("/screenshot "):
            target = user[12:].strip()
            result = screenshot(target)
            info(result)
            if "SCREENSHOT:" in result:
                path = result.split("SCREENSHOT:")[1].split("\n")[0].strip()
                pending_images.append(Path(path))
                info("Screenshot attached to next message.")
            continue

        if lo.startswith("/save"):
            parts = user.split(None, 1)
            filename = parts[1].strip() if len(parts) > 1 else f"session_{datetime.now().strftime('%Y%m%d_%H%M%S')}.md"
            if not filename.endswith(".md"):
                filename += ".md"
            try:
                lines = [f"# Agent Session — {datetime.now().strftime('%Y-%m-%d %H:%M')}\n"]
                lines.append(f"**Model:** {MODEL}  |  **CWD:** {CWD[0]}\n\n---\n")
                for msg in chat_history:
                    role = msg.get("role", "?")
                    content = msg.get("content", "")
                    if role == "system":
                        continue
                    if role == "user" and content.startswith("[Tool result:"):
                        lines.append(f"> {content[:200]}\n\n")
                    elif role == "user":
                        lines.append(f"### 👤 User\n{content}\n\n")
                    elif role == "assistant":
                        try:
                            d = json.loads(content)
                            if "final" in d:
                                lines.append(f"### 🤖 Agent\n{d['final']}\n\n")
                            elif "tool" in d:
                                args_str = json.dumps(d.get("args", {}), ensure_ascii=False)[:200]
                                lines.append(f"**Tool:** `{d['tool']}({args_str})`\n\n")
                        except Exception:
                            lines.append(f"### 🤖 Agent\n{content[:500]}\n\n")
                Path(filename).write_text("".join(lines), encoding="utf-8")
                ok(f"Session saved to {filename}")
            except Exception as e:
                err(f"Could not save: {e}")
            continue

        if lo == "/log":
            if Path(LOG_FILE).exists():
                for line in Path(LOG_FILE).read_text().splitlines()[-10:]:
                    try:
                        e = json.loads(line)
                        print(f"  Step {e.get('step', '?')} │ {e.get('tool', e.get('final', '?'))[:60]}")
                    except:
                        print(f"  {line[:60]}")
            else:
                info("No log yet.")
            continue

        if lo == "continue":
            if not context:
                warn("No paused session.")
                continue
            user = "Continue the previous task from where you left off. Review the context and proceed."

        imgs = pending_images.copy()
        pending_images.clear()

        try:
            answer, context, chat_history = run_agent(user, context, chat_history, imgs)
        except CancelledError:
            warn("Task cancelled. Session saved. You can continue later.")
            continue

        print()
        if answer == "PAUSED":
            warn("Paused. Type 'continue' to resume, or start a new task.")
        else:
            print(clr(C.BOLD + C.GREEN, "  NOVA:"))
            for line in answer.splitlines():
                print(f"  {line}")
        print()

if __name__ == "__main__":
    main()
